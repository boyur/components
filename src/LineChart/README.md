# LineChart

[Live](https://urbica.github.io/components/?selectedKind=Charts&selectedStory=LineChart)

## Usage

```js
import { LineChart } from '@urbica/components';
```

## Properties

* width: chart width **required**
* height: chart height **required**
* data: array of objects **required**
* x: x accessor function **required**
* y: y accessor function **required**
* xScale: xScale function
* yScale: yScale function
* margins: `{ top, right, bottom, left }`
* lineStyle: actual line style object


## Example

```js
import { LineChart } from '@urbica/components';
import { curveBasis } from 'd3-shape';
import { scaleLinear, scaleTime } from 'd3-scale';

<LineChart
  data={data}
  x={d => d.x}
  y={d => d.y}
  width={640}
  height={240}
  curve={curveBasis}
  xScale={scaleTime}
  yScale={scaleLinear}
  lineStyle={lineStyle}
/>
```
