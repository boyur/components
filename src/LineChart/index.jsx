import React from 'react';
import PropTypes from 'prop-types';
import { extent } from 'd3-array';
import { scaleLinear } from 'd3-scale';

import Line from '../Line';
import Chart from '../Chart';

const LineChart = (props) => {
  const { data, curve, defined, margins, lineStyle, x, y } = props;

  const width = props.width - margins.left - margins.right;
  const height = props.height - margins.top - margins.bottom;

  const xDomain = extent(data, x);
  const yDomain = extent(data, y);

  const xScale = props.xScale().range([0, width]).domain(xDomain);
  const yScale = props.yScale().range([height, 0]).domain(yDomain);

  return (
    <Chart height={props.height} width={props.width} margins={margins}>
      <Line
        data={data}
        x={x}
        y={y}
        curve={curve}
        style={lineStyle}
        xScale={xScale}
        yScale={yScale}
        defined={defined}
      />
    </Chart>
  );
};

LineChart.propTypes = {
  data: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  x: PropTypes.func.isRequired,
  y: PropTypes.func.isRequired,
  xScale: PropTypes.func,
  yScale: PropTypes.func,
  defined: PropTypes.func,
  curve: PropTypes.func,
  margins: PropTypes.shape({
    left: PropTypes.number,
    right: PropTypes.number,
    top: PropTypes.number,
    bottom: PropTypes.number
  }),
  lineStyle: PropTypes.shape({
    stroke: PropTypes.string,
    strokeWidth: PropTypes.number,
    strokeOpacity: PropTypes.number,
    strokeLinecap: PropTypes.string,
    strokeLinejoin: PropTypes.string
  }),
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired
};

LineChart.defaultProps = {
  lineStyle: {
    stroke: '#000'
  },
  margins: {
    top: 0,
    right: 0,
    bottom: 0,
    left: 0
  },
  xScale: scaleLinear,
  yScale: scaleLinear,
  curve: PropTypes.null,
  defined: PropTypes.null
};

export default LineChart;
