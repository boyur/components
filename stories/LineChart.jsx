/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import { scaleLinear, scaleTime } from 'd3-scale';
import { color, number, select } from '@storybook/addon-knobs';
import { LineChart } from '../src';

const data = [];
for (let i = 1; i <= 100; i += 1) {
  const x = new Date(3600 * 24 * 1000 * i);
  const y = Math.random() * i;
  data.push({ x, y });
}

export default () => {
  const opacityOptions = { range: true, min: 0, max: 1, step: 0.1 };
  const linecapOptions = { butt: 'butt', round: 'round', square: 'square' };
  const linejoinOptions = { miter: 'miter', round: 'round', bevel: 'bevel' };

  const lineStyle = {
    stroke: color('Line color', '#000'),
    strokeWidth: number('Line stroke width', 1),
    strokeOpacity: number('Line stroke opacity', 1, opacityOptions),
    strokeLinecap: select('Line line cap', linecapOptions, 'round'),
    strokeLinejoin: select('Line line join', linejoinOptions, 'round')
  };

  return (
    <LineChart
      data={data}
      x={d => d.x}
      y={d => d.y}
      width={number('Width', 640)}
      height={number('Height', 240)}
      xScale={scaleTime}
      yScale={scaleLinear}
      lineStyle={lineStyle}
    />
  );
};
