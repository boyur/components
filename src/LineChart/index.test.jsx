/* eslint-disable import/no-extraneous-dependencies */

import React from 'react';
import renderer from 'react-test-renderer';
import { scaleLinear } from 'd3-scale';

import LineChart from './';

test('LineChart', () => {
  const data = [{ x: 0, y: 0 }, { x: 1, y: 1 }];
  const x = d => d.x;
  const y = d => d.y;
  const style = { stroke: '#000' };

  const component = renderer.create(
    <LineChart
      data={data}
      x={x}
      y={y}
      xScale={scaleLinear}
      yScale={scaleLinear}
      style={style}
      width={100}
      height={100}
    />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
