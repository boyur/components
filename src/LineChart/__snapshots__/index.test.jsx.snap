// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LineChart 1`] = `
<svg
  height={100}
  width={100}
>
  <g
    transform="translate(0, 0)"
  >
    <path
      d="M0,100L100,0"
      fill="none"
      style={
        Object {
          "stroke": "#000",
        }
      }
    />
  </g>
</svg>
`;
